# 01. Importing libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [4]:
path = r'C:\Users\admin\06-2024 Instacart Basket Analysis'

In [5]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 03. Data consistency checks (for df_prods)

In [7]:
# Investigating the accuracy of the columns
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

## Finding and addressing mixed data types

In [9]:
# Create a dataframe
df_test = pd.DataFrame()

In [10]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [13]:
df_test['mix'] = df_test['mix'].astype('str')

## Finding and addressing missing values

In [15]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
df_prods.shape

(49693, 5)

In [19]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
df_prods_clean.shape

(49677, 5)

## Finding and addressing duplicate records

In [22]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
df_prods_clean.shape

(49677, 5)

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

# 04. Task: Data consistency checks for df_ords

In [28]:
# 2. Run the df.describe() function
df_ords.describe()

Unnamed: 0      order_id       user_id  number_of_order  \
count  3.421083e+06  3.421083e+06  3.421083e+06     3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05     1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04     1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00     1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04     5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05     1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05     2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05     1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [30]:
# 3. Check for mixed type data
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

4. I did't find any column with mixed-type data from the previous code.

In [32]:
# 5. Run a check for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
number_of_order                0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [33]:
df_ords.shape

(3421083, 8)

In [34]:
df_missingords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [35]:
df_missingords

Unnamed: 0  order_id  user_id eval_set  number_of_order  \
0                 0   2539329        1    prior                1   
11               11   2168274        2    prior                1   
26               26   1374495        3    prior                1   
39               39   3343014        4    prior                1   
45               45   2717275        5    prior                1   
...             ...       ...      ...      ...              ...   
3420930     3420930    969311   206205    prior                1   
3420934     3420934   3189322   206206    prior                1   
3421002     3421002   2166133   206207    prior                1   
3421019     3421019   2227043   206208    prior                1   
3421069     3421069   3154581   206209    prior                1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

5. I found that in the column 'days_since_prior_order', there are 206209 missing values. It could be that for the first order of any user, there wouldn't be a prior order to calculate the number of days since. Thus, it makes sense for this value to be missing or undefined.

In [37]:
# 6. Address the missing values
df_ords.fillna({"days_since_prior_order": 11}, inplace=True)

In [38]:
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
eval_set                  0
number_of_order           0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [39]:
# 6. Address the missing values
df_ords["days_since_prior_order"]

0          11.0
1          15.0
2          21.0
3          29.0
4          28.0
           ... 
3421078    29.0
3421079    30.0
3421080    18.0
3421081     7.0
3421082    30.0
Name: days_since_prior_order, Length: 3421083, dtype: float64

Note: the method taught in the exercise is no longer accepted in the new version of Python so I used the new method suggested in the instructions.

6. I chose to impute the missing data with 11 (mean = 11.111484) in days_since_prior_order. In df_prods, there were only 16 missing values, which is rather small and it's okay to just delete them. But in df_ords, 206209 rows are rather large and if I delete them, I would lose a big amount of valuable data.

In [42]:
# 7. Run a check for duplicate values
df_ords_dups = df_ords[df_ords.duplicated()]

In [43]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, number_of_order, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [44]:
df_ords.shape

(3421083, 8)

7. I didn't find any duplicates in the order data set.

8. Address duplicates: no duplicates were found. If there were, I would use df_ords.drop_duplicates() to address them.

In [47]:
# 9. Export final data
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled_clean.csv'))
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_clean.csv'))